### ISSUE TITLE
Investigate ZST-slices

### ISSUE THREAD

**Issue by oberien** [oberien/str-concat#5]
How does `concat_slice(&[(); isize::max_value()][..], &[(); isize::max_value()][..])` work? How should it be handled? We should add a comment somewhere once we know what it does / have decided on what to do.
@HeroicKatora

**Comment by 197g** [comment]
It should be ~~completely~~ mostly safe to simply pretend to have extended the first slice by the length of the second (edit: if it is initially non-empty). All invariants required seem satisfied:

* the region (which has no length) is valid for the lifetime of the result.
* the pointer of `a.as_ptr()` is non-null and aligned by construction
* the length in bytes is smaller than `isize::max_value()`

For a ZST we can no longer assume that the length is `<= isize::max_value()` and so their summed length may overflow `usize`. But we ~~don't~~ may not need the slices to be adjacent to lengthen either one to the total length. This may make the result confusing if the method works even for slices that are not located at the same address but not UB.

**Comment by 197g** [comment]
Interesting, the compiler does not like it though: [play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=88e29912f9f23dc5333d55dd288f253b)

**Comment by 197g** [comment]
Conjuring the slice of such a length from `unsafe` works fine for the rest though:

```
static S: &() = &();

fn main() {
    let s: &[()] = unsafe {
        core::slice::from_raw_parts(S, usize::max_value())
    };
    println!("{:?}", s[usize::max_value() - 1]);
}
```

[play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=81fa80ea7db55fec99551e85a843fb06)

**Comment by 197g** [comment]
Oh, using a `Vec<()>` and `resize_with`/`with_capacity` works fine as well but only in Release as it otherwise probably iterates till the end of days. This means it is actually *safe* to create slices with the length of `usize::max_value()` for at least some ZST. (And I'm left wondering if `slice::repeat` will be properly optimized here. edit: [yes](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=709167b44d5a1acf884325edb27695cf)).

```
fn main() {
    let mut s: Vec<()> = Vec::with_capacity(usize::max_value());
    s.resize_with(usize::max_value(), <()>::default);
    println!("{:?}", s[usize::max_value() - 1]);
}
```

[play](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=62f8df47bb1e739adf33338316cef27d)

**Comment by 197g** [comment]
This leaves open the question whether the new slice is allowed to point to `a.len() + b.len()` instances of `T`. **I think** this is the case when either slice has a length different from `0`, there is one instance that can be referenced and all other refer to the same instance. When neither has a length different from `0` than the result does not either.

**Comment by 197g** [comment]
Even if memory safe, concatenating ZST slices is outright weird.

```
let a: &[()] = [(); 4];
// The difference between these can not be observed within `concat_slice`:
concat_slice(&a[..2], &a[2..]);
concat_slice(&a[..2], &a[..2]);
```

Meaning, even if we were to restrict concatenating ZST slices to `a.as_ptr() == b.as_ptr()` one could generate arbitrarily long slices from a single input slice of length `>= 1`.

**Comment by 197g** [comment]
As pointed out to me [in this thread](https://github.com/rust-lang/unsafe-code-guidelines/issues/93) it may be invalid to create references to an uninhabited ZST. In that case, we must ensure *not* to create a slice of length `> 0` out of the pointer from a slice of length `0` since the latter could be valid even when the former is not. This aligns with [the comment above](https://github.com/oberien/str-concat/issues/5#issuecomment-512042125) of choose an input slice of non-zero length when the output requires it to not create references from nothing.

**Comment by oberien** [comment]
[playpen](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=264509de64041b36ba34afc69f41a70c)

Arrays, vectors and slices have special handling of ZSTs. For example a Vec of ZSTs always points to `0x1`. Multiple arrays point to their respective local stack location, but they are merged in release, resulting in different pointers on debug and the same pointer on release.

**Comment by 197g** [comment]
Iirc, *allocated* slices (vec and boxed, etc.) will point to [`NonNull::dangling`](`https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.dangling`) (which is the smallest, non-null and aligned pointer for the type, literally `mem::align_of::<T>() as *mut T`) as will any allocation of size 0 [e.g. as here for Vec](https://github.com/rust-lang/rust/blob/d56128d2919132aceaf74cc3c68a4554f5445fce/src/liballoc/raw_vec.rs#L87-L89) but that is of course not guaranteed but merely convention.

**Comment by 197g** [comment]
This question got answered: 

https://github.com/rust-lang/unsafe-code-guidelines/issues/168#issuecomment-512739217

> [..] I am sure the function presented here is safe. No need to make the question more complicated. :)

According to the definition of aliasing rules and because shared references are `Copy` it is totally safe to duplicate them by elongating a *non-mutable* slice reference *provided* it has initially been already non-empty. And since they are not allocated in any memory anyways, they are also the only types where no `unsafe` attribute of the function itself is necessary.

**Comment by oberien** [comment]
Fixed by #7. Added documentation in #9.
